Pump to 1.0 release

doctorseus opened the pull request:
These commits align the meta layer with the latest release.

- Fixed missing version string (because of missing dpkg-parsechangelog during the build process)
- Extracted zigbee plugins into its own recipe.
- https://github.com/doctorseus/meta-nymea/commit/d4fcc25a2f04b054c6a030007c54e1ed5dfb7cec should go to upstream in a later release I believe.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nymea/meta-nymea?pullRequest=23) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nymea/meta-nymea?pullRequest=23) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nymea/meta-nymea?pullRequest=23) it.</sub>

QSchulz commented on recipes-nymea/libnymea-gpio/libnymea-gpio.bb:
```diff
@@ -0,0 +1,14 @@
+DESCRIPTION = "libnymea-gpio package"
+
+LICENSE = "CLOSED"
```
It's not closed. libnymea-gpio is LGPL-3.0 | NYMEA-COMMERCIAL.

Also, technically this repository has a lib but also a tool. I would personally name the package nymea-gpio. Then have two packages, one for the lib, one for the tool.

Each will need its own license because the tools are GPL-3.0 | NYMEA-COMMERCIAL.

doctorseus replied:
Thanks for taking a look @QSchulz , I will fix this!

doctorseus replied:
Renamed the recipe and split the tool into an additional package (with a different license).

QSchulz commented on recipes-nymea/nymea-zigbee/files/0001-Make-qca-optional-only-required-for-TI-backend.patch:
```diff
@@ -0,0 +1,55 @@
+From: Lukas Pfeifhofer <lukas.pfeifhofer@streamunlimited.com>
+Date: Mon, 7 Feb 2022 10:51:34 +0100
+Subject: [PATCH 1/1] Make qca optional, only required for TI backend.
+
```
You could send this patch to nymea-zigbee git repo so that the layer does not need to carry this patch for too long (next upgrade it'd be gone).

doctorseus replied:
Yes planning to send this upstream.

doctorseus replied:
Sent: https://github.com/nymea/nymea-zigbee/pull/51

QSchulz commented on recipes-nymea/nymea-plugins/nymea-plugins_git.bb:
```diff
@@ -49,4 +49,4 @@ ALLOW_EMPTY_${PN} = "1"
 
 # Dynamically generate packages for all enabled plugins
-PACKAGES_DYNAMIC = "^nymea-plugin-.*"
+PACKAGES_DYNAMIC = "^nymea-plugin-(?!zigbee).*"
```
not sure what warrants this change? what are you trying to do here?

doctorseus replied:
So how understood it is: We have now two plugin repos, one without zigbee plugins (nymea-plugins) and one with only (nymea-plugins-zigbee).

I created the second recipe to provide the zigbee plugins in the same fashion as the non-zigbee plugins, just copied the recipe. However, now I think what is happening is that `nymea-plugins_git.bb` can not provide e.g. `nymea-plugin-zigbeegeneric` although `PACKAGES_DYNAMIC` does indicate that it would.

And thats why I run into:
```
ERROR: yocto-.... do_rootfs: Unable to install packages. Command '...' returned 255:
Collected errors:
 * opkg_prepare_url_for_install: Couldn't find anything to satisfy 'nymea-plugin-zigbeegeneric'.
``´

Although recipe `nymea-plugins-zigbee_git.bb` would indeed provide these packages.
So my way of working around this was to specify that this recipe provides everything except zigbee plugins. And then the zigbee recipe provides all zigbee ones.

Not sure if this can be done in a better way.

QSchulz replied:
> So how understood it is: We have now two plugin repos, one without zigbee plugins (nymea-plugins) and one with only (nymea-plugins-zigbee).
> 
> I created the second recipe to provide the zigbee plugins in the same fashion as the non-zigbee plugins, just copied the recipe. However, now I think what is happening is that `nymea-plugins_git.bb` can not provide e.g. `nymea-plugin-zigbeegeneric` although `PACKAGES_DYNAMIC` does indicate that it would.
> 
> And thats why I run into:
> 
> ```
> ERROR: yocto-.... do_rootfs: Unable to install packages. Command '...' returned 255:
> Collected errors:
>  * opkg_prepare_url_for_install: Couldn't find anything to satisfy 'nymea-plugin-zigbeegeneric'.
> ```
> 
> Although recipe `nymea-plugins-zigbee_git.bb` would indeed provide these packages. So my way of working around this was to specify that this recipe provides everything except zigbee plugins. And then the zigbee recipe provides all zigbee ones.
> 
> Not sure if this can be done in a better way.

Oooooh. New stuff to dig into for me as I was not aware of this mechanism! Thanks for explaining (would be a great commit log and/or comment!). You could rename the nymea-plugins plugins into something like nymea-plugin-core- or whatever suits best so that there is no naming collision.

It's a bit confusing though to have nymea-plugins and nymea-plugins-zigbee.. @mzanetti what's the rationale behind those additional git repos, seemingly storing the same thing - plugins? Is it akin to gstreamer good/bad/ugly plugins system? nymea-plugins only has first citizen plugins and then other git repos are here until the plugins are deemed stable enough to be merged into nymea-plugins?

Otherwise, I don't see another solution unfortunately (except not using the dynamic package feature which I personally really like for not having to maintain a list of PACKAGES in the recipe, which makes updates that much easier to deal with).

> Just to add some information: There are even more plugin repositories (nymea-plugins-genericthings, nymea-plugins-modbus etc...). Only the main plugins repository and the zigbee one are currently of interest to this particular project tho but this "problem" will arise more often in the future as more different plugin repositories are added.
> 
> My (somewhat underqualified) idea was to perhaps generate this from the PACKAGECONFIG which should have a list of all the built plugins of a particular repository/recipe?

That could be done but requires more active maintenance. You would need to maintain the list of PACKAGECONFIG in sync with the update, which hasn't been done in the past and is time consuming and error prone. The nice thing about the current mechanism is that any release with more/less plugins will just build the packages without any more changes than the SRCREV, create a package for it and if the user is using nymea-plugins-zigbee meta package in their image, it'll bring those new plugin packages in without any additional effort.

doctorseus replied:
As we don't really have a good alternative way of doing this I decided to keep it as it is. We will have to add new repos/plugin groups when needed, however not maintaining any kind of list of plugins is preferred.

QSchulz commented on recipes-nymea/nymea-plugins-zigbee/nymea-plugins-zigbee_git.bb:
```diff
@@ -0,0 +1,34 @@
+    nymea_libdir = d.expand('${libdir}/nymea/plugins/')
+    plugins = do_split_packages(d, nymea_libdir, r'^libnymea_integrationplugin(.*)\.so\.*', 'nymea-plugin-%s', 'Nymea integration plugin for %s', extra_depends='')
+
+    # Make nymea-plugins a meta package which RDEPENDS on all available nymea-plugin-
```
"Make nyma-plugins-zigbee a meta package [...]"

QSchulz replied:
Should have provided more context sorry.

The comment is incorrect, it should say "Make nymea-plugins-zigbee a meta package which RDEPENDS on all available nymea-plugin-zigbee*"

This is due to ${PN} being used which is nymea-plugins-zigbee in that recipe and not nymea-plugins, the package recipe you most likely copy pasted this from :)

doctorseus replied:
Fixed.

QSchulz commented on recipes-nymea/nymea-plugins-zigbee/nymea-plugins-zigbee_git.bb:
```diff
@@ -0,0 +1,34 @@
+
+python split_nymea_plugins_packages() {
+    nymea_libdir = d.expand('${libdir}/nymea/plugins/')
+    plugins = do_split_packages(d, nymea_libdir, r'^libnymea_integrationplugin(.*)\.so\.*', 'nymea-plugin-%s', 'Nymea integration plugin for %s', extra_depends='')
```
could add zigbee outside of the matching pattern and have "Nymea Zigbee integration for %s" as text. But that's nitpick, it's probably just fine like this.

doctorseus replied:
For this I decided to keep this as it is for now.

QSchulz commented on recipes-nymea/nymea-plugins-zigbee/nymea-plugins-zigbee_git.bb:
```diff
@@ -0,0 +1,34 @@
+
+S = "${WORKDIR}/git"
+
+EXTRA_QMAKEVARS_PRE += "CONFIG+=selection ${PACKAGECONFIG_CONFARGS}"
```
I think this is currently a no-op so can be removed.

mzanetti replied:
Oh, also, all the repositories have new build options now:
`WITH_PLUGINS="plugin1 plugin2 ..." `
would explicitly list plugins to be included (pretty much how the old `CONFIG+=selection SELECTION="plugin1 plugin2"` did and
`WITHOUT_PLUGINS="plugin1 plugin2 ..."`
would build all but the listed.

In any case, the "CONFIG+=selection" is no more...

mzanetti replied:
Some of the devices do not follow the zigbee specification and have a reverse engineered protocol. While they work perfectly fine and are very unlikely to stop working with a firmware update, it may still be desirable to disable them when building a product and only supply those who officially want to be connectable by 3rd party devices.

doctorseus replied:
Fixed and removed CONFIG+selection part.

QSchulz reviewed (changes requested):
Some small comments to address :) Mostly looks good otherwise (haven't checked the hashes though).

doctorseus commented:
I fixed a few things based on the comments from @QSchulz. Thanks for that! 
Updated the branch and also changed the commit author so @CLAassistant isn't confused again. ;-)
